homeassistant/components/tomorrowio/__init__.py

```diff
@@ -227,54 +217,105 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
+                self.add_entry_to_location_dict(entry_)
+            await super().async_config_entry_first_refresh()
+            self._coordinator_ready.set()
+        else:
+            # If we have an event, we need to wait for it to be set before we proceed
+            await self._coordinator_ready.wait()
+            # If we're loading a new config entry that's not already mapped, we need
+            # to do a refresh. We're going to do a partial refresh though so we can
+            # minimize repeat API calls
+            if entry.entry_id not in self.entry_id_to_location_dict:
+                self.add_entry_to_location_dict(entry)
+                await super().async_refresh()
+            # If we're not getting new data, we don't need to schedule a refresh
+            else:
+                return
+
+        self.update_interval = async_set_update_interval(self.hass, self._api)
+        self._schedule_refresh()
+
+    async def async_unload_entry(self, entry: ConfigEntry) -> bool | None:
+        """
+        Unload a config entry from coordinator.
 
-        super().__init__(
-            hass,
-            _LOGGER,
-            name=config_entry.data[CONF_NAME],
-            update_interval=update_interval,
-        )
+        Returns whether coordinator can be removed as well because there are no
+        config entries tied to it anymore.
+        """
+        self.entry_id_to_location_dict.pop(entry.entry_id)
+        self.update_interval = async_set_update_interval(self.hass, self._api, entry)
+        return not self.entry_id_to_location_dict
 
     async def _async_update_data(self) -> dict[str, Any]:
         """Update data via library."""
+        data = {}
+        # If we are refreshing because of a new config entry that's not already in our
+        # data, we do a partial refresh to avoid wasted API calls.
+        if self.data and any(
+            entry_id not in self.data for entry_id in self.entry_id_to_location_dict
+        ):
+            data = self.data
+
         try:
-            return await self._api.realtime_and_all_forecasts(
-                [
-                    TMRW_ATTR_TEMPERATURE,
-                    TMRW_ATTR_HUMIDITY,
-                    TMRW_ATTR_PRESSURE,
-                    TMRW_ATTR_WIND_SPEED,
-                    TMRW_ATTR_WIND_DIRECTION,
-                    TMRW_ATTR_CONDITION,
-                    TMRW_ATTR_VISIBILITY,
-                    TMRW_ATTR_OZONE,
-                    TMRW_ATTR_WIND_GUST,
-                    TMRW_ATTR_CLOUD_COVER,
-                    TMRW_ATTR_PRECIPITATION_TYPE,
-                    *(
+            for entry_id, location in self.entry_id_to_location_dict.items():
+                if entry_id in data:
+                    continue
+                entry = self.hass.config_entries.async_get_entry(entry_id)
+                assert entry
+                data[entry_id] = await self._api.realtime_and_all_forecasts(
```

Comment:
Can we move the try... except block so that it only wraps the line that can raise?